Compilation failure

[ 36%] Building CXX object percona_playback/query_log/CMakeFiles/query_log.dir/query_log.cc.o
/opt/vadim/query/query-playback/percona_playback/query_log/query_log.cc:55:8: error: â€˜atomicâ€™ in namespace â€˜boostâ€™ does not name a type
 static boost::atomic<long long> g_max_behind_ns;
        ^
/opt/vadim/query/query-playback/percona_playback/query_log/query_log.cc: In function â€˜bool parse_time(boost::string_ref, QueryLogEntry::TimePoint&)â€™:
/opt/vadim/query/query-playback/percona_playback/query_log/query_log.cc:89:27: error: â€˜boost::string_refâ€™ has no member named â€˜to_stringâ€™
   int num_read = sscanf(s.to_string().c_str(), "# Time: %02d%02d%02d %2d:%02d:%02d.%06lld",

after https://github.com/Percona-Lab/query-playback/pull/28

I am getting:

In file included from /opt/vadim/query/query-playback/percona_playback/query_log/query_log.h:27:0,
                 from /opt/vadim/query/query-playback/percona_playback/query_log/query_log.cc:27:
/usr/include/boost/utility/string_ref.hpp: In instantiation of â€˜boost::basic_string_ref<charT, traits>::size_type boost::basic_string_ref<charT, traits>::find(boost::basic_string_ref<charT, traits>) const [with charT = char; traits = std::char_traits<char>; boost::basic_string_ref<charT, traits>::size_type = long unsigned int]â€™:
/opt/vadim/query/query-playback/percona_playback/query_log/query_log.h:63:56:   required from here
/usr/include/boost/utility/string_ref.hpp:175:25: error: invalid conversion from â€˜long unsigned intâ€™ to â€˜boost::basic_string_ref<char, std::char_traits<char> >::const_iterator {aka const char*}â€™ [-fpermissive]
             return iter = this->cend () ? npos : std::distance ( this->cbegin (), iter );
                         ^
/usr/include/boost/utility/string_ref.hpp:175:88: error: invalid conversion from â€˜boost::basic_string_ref<char, std::char_traits<char> >::const_iterator {aka const char*}â€™ to â€˜boost::basic_string_ref<char, std::char_traits<char> >::size_type {aka long unsigned int}â€™ [-fpermissive]
             return iter = this->cend () ? npos : std::distance ( this->cbegin (), iter );
                                                                                        ^
make[2]: *** [percona_playback/query_log/CMakeFiles/query_log.dir/query_log.cc.o] Error 1
make[1]: *** [percona_playback/query_log/CMakeFiles/query_log.dir/all] Error 2

--- a/README.md
+++ b/README.md
@@ -20,7 +20,7 @@ You will need the 'development' packages for all of these (often suffixed by
 The core of Query Playback requires:
 
  * libtbb (Intel Threading Building Blocks)
- * Boost >= 1.52
+ * Boost >= 1.53
  * pkg-config
  * cmake >= 2.8.7
 

--- a/percona_playback/CMakeLists.txt
+++ b/percona_playback/CMakeLists.txt
@@ -27,7 +27,7 @@ include_directories(${PROJECT_SOURCE_DIR})
 include_directories(${CMAKE_CURRENT_BINARY_DIR})
 
 # setup db_thread lib which all plugins depend on
-find_package(Boost COMPONENTS system thread regex program_options chrono REQUIRED)
+find_package(Boost 1.53.0 COMPONENTS system thread regex program_options chrono REQUIRED)
 include_directories(${Boost_INCLUDE_DIR})
 link_directories(${Boost_LIBRARY_DIR})
 add_library(db_thread STATIC db_thread.cc)

--- a/percona_playback/mysql_client/CMakeLists.txt
+++ b/percona_playback/mysql_client/CMakeLists.txt
@@ -24,10 +24,13 @@
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/percona_playback")
 include(plugin)
 
-find_path(MYSQL_INC "mysql/mysql.h")
-include_directories("${MYSQL_INC}/mysql")
+find_file(MYSQL_INC "mysql.h" PATH_SUFFIXES "mysql")
+get_filename_component(MYSQL_INC ${MYSQL_INC} PATH)
+include_directories(${MYSQL_INC})
+
+find_library(MYSQL_LIB "mysqlclient_r" PATH_SUFFIXES "mysql")
 
 add_library(mysql_client mysql_client.cc)
-target_link_libraries(mysql_client mysqlclient_r crypto ssl)
+target_link_libraries(mysql_client ${MYSQL_LIB} crypto ssl)
 
 setup_plugin("mysql_client")

--- a/percona_playback/query_log/query_log.cc
+++ b/percona_playback/query_log/query_log.cc
@@ -57,6 +57,13 @@ static boost::atomic<long long> g_max_behind_ns;
 
 extern percona_playback::DispatcherPlugin *g_dispatcher_plugin;
 
+// workaround bug in boost 1.53 string_ref::find(string_ref)
+static boost::string_ref::size_type find(boost::string_ref str, boost::string_ref substr) {
+  boost::string_ref::const_iterator iter;
+  iter = std::search(str.cbegin(), str.cend(), substr.cbegin(), substr.cend());
+  return iter == str.cend() ? boost::string_ref::npos : std::distance(str.cbegin(), iter);
+}
+
 // returns next line starting from current position inside the data and updating it.
 // a empty string signals that the end is reached.
 static boost::string_ref readline(boost::string_ref data, boost::string_ref::size_type& pos) {
@@ -253,6 +260,10 @@ void QueryLogEntry::execute(DBThread *t)
   }
 }
 
+bool QueryLogEntry::is_quit() const {
+  return find(data, "\n# administrator command: Quit;") != boost::string_ref::npos;
+}
+
 std::string QueryLogEntry::getQuery(bool remove_timestamp) {
   std::string ret;
   boost::string_ref::size_type pos = 0;
@@ -275,33 +286,33 @@ std::string QueryLogEntry::getQuery(bool remove_timestamp) {
 }
 
 uint64_t QueryLogEntry::parseThreadId() const {
-  size_t location= data.find("Thread_id: ");
+  size_t location= find(data, "Thread_id: ");
   if (location != std::string::npos)
     return strtoull(&data[location + strlen("Thread_Id: ")], NULL, 10);
 
   // starting from MySQL 5.6.2 (bug #53630) the thread id is included as "Id:"
-  location= data.find("Id: ");
+  location= find(data, "Id: ");
   if (location != std::string::npos)
     return strtoull(&data[location + strlen("Id: ")], NULL, 10);
   return 0;
 }
 
 uint64_t QueryLogEntry::parseRowsSent() const {
-  size_t location= data.find("Rows_sent: ");
+  size_t location= find(data, "Rows_sent: ");
   if (location != std::string::npos)
     return strtoull(&data[location + strlen("Rows_sent: ")], NULL, 10);
   return 0;
 }
 
 uint64_t QueryLogEntry::parseRowsExamined() const {
-  size_t location= data.find("Rows_Examined: ");
+  size_t location= find(data, "Rows_Examined: ");
   if (location != std::string::npos)
     return strtoull(&data[location + strlen("Rows_examined: ")], NULL, 10);
   return 0;
 }
 
 double QueryLogEntry::parseQueryTime() const {
-  size_t location= data.find("Query_time: ");
+  size_t location= find(data, "Query_time: ");
   if (location != std::string::npos)
     return strtod(&data[location + strlen("Query_time: ")], NULL);
   return 0.0;

--- a/percona_playback/query_log/query_log.h
+++ b/percona_playback/query_log/query_log.h
@@ -58,10 +58,7 @@ public:
 
   virtual void execute(DBThread *t);
 
-  virtual bool is_quit() const
-  {
-    return data.find("\n# administrator command: Quit;") != boost::string_ref::npos;
-  }
+  virtual bool is_quit() const;
 
   uint64_t parseThreadId() const;
   uint64_t parseRowsSent() const;

